Added cmake-compile command, changed prefixing

Tyler-g-hudson opened the pull request:
This PR is slightly small so I decided to add a few 

+ Added the cmake-compile command and all associated functionality
+ Added tag prefixing for base tags as well as image tags to make the interface more consistent.
~ Changed tag prefixing to use a function in most cases
~ Changed no-cache to be handled by a no-cache parser in the cli
~ Small naming convention changes to be more consistent

Tyler-g-hudson committed:
Added cmake-compile command, changed prefixing

+ Added the cmake-compile command and all associated functionality
~ Changed tag prefixing to use a function in most cases
+ Added tag prefixing for base tags as well as image tags to make the interface more consistent.

gmgunter commented on docker_cli/_docker_cmake.py:
```diff
@@ -60,3 +60,29 @@ def cmake_config_dockerfile(base: str, build_type: str, with_cuda: bool = True)
+    # Run as the $MAMBA_USER and activate the micromamba environment.
+    dockerfile += f"{micromamba_docker_lines()}\n\n"
+
+    # build the project.
```
Elsewhere in this function, the comments all start with a capital letter, so let's try to be consistent.

```suggestion
    # Build the project.
```

gmgunter reviewed (commented):
Looks good, thanks.

We're just missing a unit test for the new functionality. The simplest (and probably best) test would be to build isce3 inside the container (via `cmake_build_dockerfile()`) and then check the container for the existence of the following file in the build directory: `cxx/isce3/libisce3.so`.

The only issue with that test is that it'll be pretty slow since it takes a while to build isce3. So we should probably mark the test as "slow" using [pytest markers](https://docs.pytest.org/en/latest/how-to/mark.html#registering-marks). Another approach would be to try and find a simple C++ project (that uses CMake) which we could clone into the Docker container and build instead of isce3 as a faster test.

Tyler-g-hudson committed:
Tiny documentation change.

Tyler-g-hudson committed:
setup_conda_dev bugfix

~ Removed a bug that causes errors when building the conda dev image.

Tyler-g-hudson committed:
Lock file update

~ Updated the lock files.

Tyler-g-hudson committed:
Bugfixes

~ Updated the GCC version in the lockfiles
~ Added a Test option for docker init images to avoid unintended removals

Tyler-g-hudson committed:
Added ISCE3 test fixtures and CMake build test.

gmgunter commented on docker_cli/_docker_init.py:
```diff
@@ -11,13 +11,16 @@ def init_dockerfile(base: str, custom_lines: str) -> str:
         The name of the image upon this one will be based.
     custom_lines : str
         Custom installation and configuration lines to be added to the Dockerfile.
+    no_cache : bool, optional
```
```suggestion
    test : bool, optional
```

gmgunter commented on docker_cli/setup_commands.py:
```diff
@@ -28,6 +27,9 @@ def setup_init(
         The tag of the image to be built.
     no_cache : bool
         Run Docker build with no cache if True.
+    no_cache : bool, optional
```
```suggestion
    test : bool, optional
```

gmgunter commented on docker_cli/setup_commands.py:
```diff
@@ -320,26 +322,30 @@ def setup_all(
         The location of the dev environment requirements file.
     verbose : bool, optional
         If True, output informational messages upon completion. Defaults to False.
+        Add a "test" directory to the init image if True. Used for test images.
```
```suggestion
    test : bool, optional
        Add a "test" directory to the init image if True. Used for test images.
```

gmgunter commented on test/fixtures_isce3.py:
```diff
@@ -0,0 +1,171 @@
+    """Return the tag of the ISCE3 development environment image."""
+    pattern = re.compile(rf".*{isce3_build_tag}.*mamba-dev")
+
+    env_dev_image_tag: str = ""
+
+    for tag in isce3_setup_images:
+        if pattern.match(tag):
+            env_dev_image_tag = tag
+
+    if not env_dev_image_tag:
+        raise ValueError("No development environment image tag found.")
+
+    return env_dev_image_tag
```
Let's return as soon as we find the first tag that matches `pattern`. No need to continue looping over the remaining image tags after that.

```suggestion
    for tag in isce3_setup_images:
        if pattern.match(tag):
            return tag

    raise ValueError("No development environment image tag found.")
```

gmgunter commented on test/fixtures_isce3.py:
```diff
@@ -0,0 +1,171 @@
+    )
+
+    yield Image.build(
+        tag=isce3_git_repo_tag, dockerfile_string=dockerfile, no_cache=True
```
Are we deliberately using `no_cache=True` here and for the config/build images below? I thought based on our slack convo the other day that we were OK with using the cache for these images.

gmgunter reviewed (commented):
Thanks for the updates!

Tyler-g-hudson committed:
Documentation updates

gmgunter commented on test/fixtures_isce3.py:
```diff
@@ -0,0 +1,165 @@
+    )
+
+    yield Image.build(
+        tag=isce3_cmake_config_tag, dockerfile_string=dockerfile, no_cache=True
```
Should we use the cache here as well?

```suggestion
        tag=isce3_cmake_config_tag, dockerfile_string=dockerfile, no_cache=False
```

gmgunter commented on test/fixtures_isce3.py:
```diff
@@ -0,0 +1,165 @@
+    )
+
+    yield Image.build(
+        tag=isce3_cmake_build_tag, dockerfile_string=dockerfile, no_cache=True
```
```suggestion
        tag=isce3_cmake_build_tag, dockerfile_string=dockerfile, no_cache=False
```

gmgunter reviewed (approved):
LGTM, thanks!

Tyler-g-hudson committed:
ISCE3 fixtures: No-Cache -> False
